infra/gravity/provision.go

```diff
@@ -189,99 +190,98 @@ Loop:
-	ctx, cancel = context.WithTimeout(c.Context(), diskWaitTimeout)
+	ctx, cancel := context.WithTimeout(c.Context(), diskWaitTimeout)
 	defer cancel()
-	err = waitDisks(ctx, gravityNodes, []string{"/iotest", path.Join(cfg.dockerDevice, "/iotest")})
+
+	err = waitDisks(ctx, gravityNodes, []string{"/iotest", path.Join(cfg.dockerDevice, "/iotest")}, c.Logger())
 	if err != nil {
 		err = trace.Wrap(err, "VM disks do not meet minimum write performance requirements")
 		c.Logger().WithError(err).Error(err.Error())
-		return nil, destroyFn, trace.Wrap(err)
+		return cluster, trace.Wrap(err)
 	}
 
+	cluster.Nodes = gravityNodes
 	c.Logger().Info("provisioning complete")
-	return gravityNodes, destroyFn, nil
+	return cluster, nil
 }
 
 // provisionCloud gets VMs up, running and ready to use
-func (c *TestContext) provisionCloud(cfg ProvisionerConfig) (gravityNodes []Gravity, destroyResources DestroyFn, err error) {
-	c.Logger().WithField("config", cfg).Debug("Provisioning VMs")
+func (c *TestContext) provisionCloud(cfg ProvisionerConfig) (cluster Cluster, config *terraform.Config, err error) {
+	log := c.Logger().WithField("config", cfg)
+	log.Debug("Provisioning VMs")
 
 	err = validateConfig(cfg)
 	if err != nil {
-		return nil, nil, trace.Wrap(err)
+		return cluster, nil, trace.Wrap(err)
 	}
 
-	nodes, destroyFn, params, err := runTerraform(c.Context(), cfg, c.Logger())
+	infra, err := runTerraform(c.Context(), cfg, c.Logger())
 	if err != nil {
-		return nil, nil, trace.Wrap(err)
+		return cluster, nil, trace.Wrap(err)
 	}
 	defer func() {
-		if err == nil {
+		if err == nil || infra.destroyFn == nil {
 			return
 		}
-		if errDestroy := destroyResource(destroyFn); errDestroy != nil {
-			c.Logger().WithError(errDestroy).Error("Failed to destroy resources.")
+		if errDestroy := destroyResource(infra.destroyFn); errDestroy != nil {
+			log.WithError(errDestroy).Error("Failed to destroy resources.")
 		}
 	}()
 
 	ctx, cancel := context.WithTimeout(c.Context(), cloudInitTimeout)
 	defer cancel()
 
-	c.Logger().Debug("Configuring VMs")
-	gravityNodes, err = configureVMs(ctx, c.Logger(), *params, nodes)
+	log.Debug("Configuring VMs")
+	gravityNodes, err := configureVMs(ctx, c.Logger(), infra.params, infra.nodes)
 	if err != nil {
-		c.Logger().WithError(err).Error("Some nodes failed to initialize, tear down as non-usable.")
-		return nil, nil, trace.Wrap(err)
+		log.WithError(err).Error("Some nodes failed to initialize, tear down as non-usable.")
+		return cluster, nil, trace.NewAggregate(err, destroyResource(infra.destroyFn))
 	}
 
 	err = c.postProvision(cfg, gravityNodes)
 	if err != nil {
-		c.Logger().WithError(err).Error("Post-provisioning failed, tear down as non-usable.")
-		return nil, nil, trace.Wrap(err)
+		log.WithError(err).Error("Post-provisioning failed, tear down as non-usable.")
+		return cluster, nil, trace.Wrap(err)
 	}
 
-	c.Logger().Debug("ensuring disk speed is adequate across nodes")
```

Comment:
This seems to have only affected `Azure` as far as I remember.

Reply:
Azure is still supported provider for robotest, right? Maybe keep this for it at least, if it's still relevant that is.